jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -0,0 +1,792 @@
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DoSHandler(
+        @Name("getId") Function<Request, String> getId,
+        @Name("trackerFactory") Tracker.Factory trackerFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        this(null, getId, trackerFactory, rejectHandler, maxTrackers);
+    }
+
+    /**
+     * @param handler Then next {@link Handler} or {@code null}
+     * @param getId Function to extract an remote ID from a request.
+     * @param trackerFactory Factory to create a Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DoSHandler(
+        @Name("handler") Handler handler,
+        @Name("getId") Function<Request, String> getId,
+        @Name("trackerFactory") Tracker.Factory trackerFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        super(handler);
+        installBean(_trackers);
+        _getId = Objects.requireNonNullElse(getId, ID_FROM_REMOTE_ADDRESS);
+        installBean(_getId);
+        _trackerFactory = Objects.requireNonNull(trackerFactory);
+        installBean(_trackerFactory);
+        _maxTrackers = maxTrackers < 0 ? 10_000 : maxTrackers;
+        _rejectHandler = Objects.requireNonNullElseGet(rejectHandler, StatusRejectHandler::new);
+        installBean(_rejectHandler);
+    }
+
+    @Override
+    public void setServer(Server server)
+    {
+        super.setServer(server);
+        if (_rejectHandler instanceof Handler handler)
+            handler.setServer(server);
+    }
+
+    @Override
+    protected boolean onConditionsMet(Request request, Response response, Callback callback) throws Exception
+    {
+        // Reject if we have too many Trackers
+        if (_maxTrackers > 0 && _trackers.size() >= _maxTrackers)
+        {
+            // Try shrinking the tracker pool
+            long now = NanoTime.now();
+            _trackers.values().removeIf(tracker -> tracker.isIdle(now));
+            if (_trackers.size() >= _maxTrackers)
+            {
+                // Try shrinking the tracker pool as if we are at the next sample period already
+                long nextIdleCheck = NanoTime.now() + _trackerFactory.getSamplePeriod().getNano();
+                _trackers.values().removeIf(tracker -> tracker.isIdle(nextIdleCheck));
+                if (_trackers.size() >= _maxTrackers)
```

Comment:
There should not be the need to check here for idleness, this is handled by `CyclicTimeouts`.

Reply:
This is an aggressive pushing of idleness to remove almost idle trackers.